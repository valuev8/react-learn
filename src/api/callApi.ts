import axios from 'axios';
import {SERVER_API_URL} from '../shared/constants/constants';

axios.interceptors.response.use((response) => response.data);

export default function callApi(endpoint: string, method = 'get', body?: any, params?: any) {
  return body ?
    axios[method](`${SERVER_API_URL}/${endpoint}`, body) :
    axios[method](`${SERVER_API_URL}/${endpoint}`, {params});
}
