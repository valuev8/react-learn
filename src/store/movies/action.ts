import {
  APPLY_FILTER,
  APPLY_SEARCH,
  APPLY_SORT,
  CREATE_MOVIE,
  DELETE_MOVIE,
  EDIT_MOVIE,
  GET_MOVIES,
  HANDLE_ERROR,
} from './types';
import {Movie} from '../../shared/models/movie.type';
import {RootState} from '../rootReducer';

export const getMovies = (payload: Movie[]) => ({
  type: GET_MOVIES,
  payload,
});

export const deleteMovie = (payload: number) => ({
  type: DELETE_MOVIE,
  payload,
});

export const editMovie = (payload: Movie) => ({
  type: EDIT_MOVIE,
  payload,
});

export const createMovie = (payload: Movie) => ({
  type: CREATE_MOVIE,
  payload,
});

// TODO: add error notification
export const handleError = (payload: Error) => ({
  type: HANDLE_ERROR,
  payload,
});

export const applyFilter = (payload: string) => ({
  type: APPLY_FILTER,
  payload,
});

export const applySort = (payload: string) => ({
  type: APPLY_SORT,
  payload,
});

export const applySearch = (payload: string) => ({
  type: APPLY_SEARCH,
  payload,
});

// TODO: Define types for dispatch and api / add loader
export const getMoviesThunk = () => (dispatch: any, state: RootState, api: any) => api('movies')
    .then((res: { data: Movie[] }) => dispatch(getMovies(res.data)))
    .catch((err: Error) => dispatch(handleError(err)));

export const deleteMovieThunk = (movieId: number) => (dispatch: any, state: RootState, api: any) =>
  api(`movies/${movieId}`, 'delete')
      .then(() => dispatch(deleteMovie(movieId)))
      .catch((err: Error) => dispatch(handleError(err)));

export const editMovieThunk = (movie: Movie) => (dispatch: any, state: RootState, api: any) =>
  api(`movies`, 'put', movie)
      .then(() => dispatch(editMovie(movie)))
      .catch((err: Error) => dispatch(handleError(err)));

export const createMovieThunk = (movie: Partial<Movie>) => (dispatch: any, state: RootState, api: any) =>
  api(`movies`, 'post', movie)
      .then((createdMovie: Movie) => dispatch(createMovie(createdMovie)))
      .catch((err: Error) => dispatch(handleError(err)));

