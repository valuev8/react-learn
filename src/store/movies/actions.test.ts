import configureMockStore from 'redux-mock-store';
import {createMovieThunk, deleteMovieThunk, editMovieThunk, getMoviesThunk} from './action';
import thunk from 'redux-thunk';
import callApi from '../../api/callApi';
import axios from 'axios';
import {Movie} from '../../shared/models/movie.type';

let store: any;
const mockMovie = {
  title: 'test',
};
jest.mock('axios');
const middlewares = [thunk.withExtraArgument(callApi)];
const mockStore = configureMockStore(middlewares);

describe('movie actions', () => {
  beforeEach(() => {
    store = mockStore({});
  });

  it('creates GET_MOVIES when getting movies was successful', () => {
    // @ts-ignore
    axios.get.mockResolvedValueOnce({data: [mockMovie]});

    return store.dispatch(getMoviesThunk())
        .then(() => {
          expect(store.getActions()).toMatchSnapshot();
        });
  });

  it('creates DELETE_MOVIES when deleting movies was successful', () => {
    // @ts-ignore
    axios.delete.mockResolvedValueOnce({data: [mockMovie]});

    return store.dispatch(deleteMovieThunk(1))
        .then(() => {
          expect(store.getActions()).toMatchSnapshot();
        });
  });

  it('creates EDIT_MOVIES when updating movies was successful', () => {
    // @ts-ignore
    axios.put.mockResolvedValueOnce({data: [mockMovie]});

    return store.dispatch(editMovieThunk(mockMovie as Movie))
        .then(() => {
          expect(store.getActions()).toMatchSnapshot();
        });
  });

  it('creates CREATE_MOVIE when creating movie was successful', () => {
    // @ts-ignore
    axios.post.mockResolvedValueOnce({data: [mockMovie]});

    return store.dispatch(createMovieThunk(mockMovie as Movie))
        .then(() => {
          expect(store.getActions()).toMatchSnapshot();
        });
  });
});
