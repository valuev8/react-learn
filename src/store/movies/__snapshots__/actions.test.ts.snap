// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`movie actions creates CREATE_MOVIE when creating movie was successful 1`] = `
Array [
  Object {
    "payload": Object {
      "data": Array [
        Object {
          "title": "test",
        },
      ],
    },
    "type": "CREATE_MOVIE",
  },
]
`;

exports[`movie actions creates DELETE_MOVIES when deleting movies was successful 1`] = `
Array [
  Object {
    "payload": 1,
    "type": "DELETE_MOVIE",
  },
]
`;

exports[`movie actions creates EDIT_MOVIES when updating movies was successful 1`] = `
Array [
  Object {
    "payload": Object {
      "title": "test",
    },
    "type": "EDIT_MOVIE",
  },
]
`;

exports[`movie actions creates GET_MOVIES when getting movies was successful 1`] = `
Array [
  Object {
    "payload": Array [
      Object {
        "title": "test",
      },
    ],
    "type": "GET_MOVIES",
  },
]
`;
